### ISSUE TITLE
c++ Frontend: example mnist hits nan

### ISSUE THREAD

**Issue by an-kumar** [pytorch/pytorch#15522]
## 🐛 Bug
The example mnist.cpp for the c++ frontend hits nan values around epoch 6.
## To Reproduce

1. Download the example here (from the docs): https://github.com/goldsborough/examples/tree/cpp/cpp
2. Compile and run
3. nans will show up around epoch 6

## Expected behavior

Successfully trains mnist. Note that before the nans hit, the accuracy is around 97-98% or so.

## Environment

 - PyTorch Version (e.g., 1.0): c++ frontend 1.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): N/A
 - Build command you used (if compiling from source): cmake && make
 - CUDA/cuDNN version: 9.0
 - GPU models and configuration: GTX 1080TI
 - Any other relevant information: Nans also show up if no cuda is used.

## Additional context

I actually investigated a decent amount here and was unable to solve the issue. I did the following:

1- update the cpp mnist to be in line with the py mnist, which has a different network now
2- Implement kaiming uniform initialization in cpp and used that instead of the default initialization (py uses kaiming uniform, so I assumed that was the issue)

When I do those things, instead of getting nans, the network still trains well at first (getting to ~98% accuracy), but at some point blows up and instead of getting nans, the loss blows up to 2.3 and all the outputs are the exact same (presumably a gradient blew up and the network's relu at the end becomes saturated)

As a separate comment, I'm curious what the interaction between the python and c++ libraries are. Are they *both* calling into the same shared underlying c++ library, or is the python library calling into the c++ frontend? Could the c++ implementation be bugged in that sense?

By the way there's a unit test running against mnist that only runs for 3 epochs. If you set the number of epochs to 10 there I would imagine it would hit this error.

Additionally, I'm not sure that the current distribution of libtorch was compiled with cudnn support. It's quite a bit slower (on the simple mnist example) than python; and if I explicitly remove `libcudnn` from the linker, it doesn't complain at all. Could be related -- perhaps cudnn is more numerically stable?

**Comment by soumith** [comment]
cc: @goldsborough

**Comment by apchenstu** [comment]
I got the same problem, the training/test accuracy is super low in my side, only 0.16.

**Comment by apchenstu** [comment]
Oh I realize there is a bug in the example code, the training dataset and test size should not be the same, so the I got a very low accuracy.
 [https://github.com/goldsborough/examples/blob/cpp/cpp/mnist/mnist.cpp#L137](url)

**Comment by goldsborough** [comment]
@apchenstu Yes I saw that bug too, it's fixed on master now. Trying to figure out the instability issue

**Comment by goldsborough** [comment]
The NaN/instability issue is now fixed. The model converges successfully.
